module-system/sov-state/src/codec/mod.rs

```diff
@@ -42,3 +43,38 @@ pub trait StateValueCodec<V> {
             .unwrap()
     }
 }
+
+pub trait EncodeLike<Ref: ?Sized, Target> {
+    fn encode_like(&self, borrowed: &Ref) -> Vec<u8>;
+}
+
+// All items can be encoded like themselves
+impl<C, T> EncodeLike<T, T> for C
+where
+    C: StateValueCodec<T>,
+{
+    fn encode_like(&self, borrowed: &T) -> Vec<u8> {
+        self.encode_value(borrowed)
+    }
+}
+
+// In borsh, a slice is encoded the same way as a vector
+impl<T> EncodeLike<[T], Vec<T>> for BorshCodec
+where
+    T: BorshSerialize,
+{
+    fn encode_like(&self, borrowed: &[T]) -> Vec<u8> {
```

Comment:
We probably shouldn't make bigger changes now - especially under the bugfix context - but an alternative that might require less cloning is if we define a serializing interface, such as:

```rust
fn encode_like<S, B>(&self, serializer: &mut S, value: B)
where
  S: Serializer,
  B: AsRef<[T]>
```

This way we avoid forcing owned vecs and reduce allocation

Reply:
(in response to @vlopes11)

The `Serializer`/`Writer` pattern is interesting, it might be a good fit here and reduce cloning and `Vec<u8>` allocations. I think we should keep it in mind for future performance optimizations.

As for `AsRef<[T]>` (why `[T]` and not `T`? That's not clear to me) we have to be really careful not to run into [these](https://github.com/Sovereign-Labs/sovereign-sdk/pull/644#discussion_r1294482223) problems, which apply to `AsRef` just as much as `Borrow`: https://doc.rust-lang.org/src/core/array/mod.rs.html#159

Reply:
This is a good idea, but let's leave it for a follow-up since it's not on the critical path.